src/components/swarm.rs

```diff
@@ -423,3 +423,127 @@ impl ParticleSwarmUpdate<Global> {
+    where
+        P: LimitedVectorProblem<Element = f64>,
+        I: Identifier,
+{
+    fn init(&self, _problem: &P, state: &mut State<P>) -> ExecResult<()> {
+        state.insert(RandomizationParameter(self.alpha));
+        Ok(())
+    }
+
+    fn execute(&self, problem: &P, state: &mut State<P>) -> ExecResult<()> {
+        let mut populations = state.populations_mut();
+        let mut rng = state.random_mut();
+
+        // Prepare parameters
+        let &Self {
+            beta, gamma, ..
+        } = self;
+        let a = state.get_value::<RandomizationParameter>();
+
+        // Get necessary state
+        let mut jj: Vec<Individual<P>> = populations.current().into_iter().cloned().collect();
+        let mut ii: Vec<Individual<P>> = populations.current().into_iter().cloned().collect();
+        let mut x: Vec<&mut Vec<f64>> = populations.current_mut().as_solutions_mut();
+
+        // scale for adapting to problem domain; at the moment only if domain in each dim is the same
+        let scale = (problem.domain()[0].end - problem.domain()[0].start).abs();
+
+        // Perform the update step.
+        // compare all individuals
+        for (u, i) in  ii.iter_mut().enumerate() {
+            for j in jj.iter_mut() {
+                // if individual j is "more attractive" (i.e. has lower fitness), move towards j
+                if i.get_objective() > j.get_objective() {
+                    // draw random values from uniform distribution between 0 and 1
+                    // according to paper: also possible to use normal distribution, depending on problem
+                    let rand: Vec<f64> = (0..problem.dimension()).map(|_| rng.gen_range(0.0..1.0)).collect();
+                    // calculate distance between i and j; without .sqrt() as it has to be squared again in the next step
+                    let r = i.solution_mut()
+                        .into_iter()
+                        .zip(j.solution_mut())
+                        .map(|(p, q)| (p.clone() - q.clone()).powf(2.0))
+                        .sum::<f64>();
+                    // calculate "attractiveness"
+                    let b = beta * (- gamma * r).exp();
+                    // calculate difference of solutions j and i
+                    let diff = i.solution_mut()
+                        .into_iter()
+                        .zip(j.solution_mut())
+                        .map(|(p, q)| q.clone() - p.clone())
+                        .collect::<Vec<f64>>();
+                    // calculate values that should be added to current position
+                    let pos = diff
+                        .into_iter()
+                        .zip(rand)
+                        .map(|(p, q)| b * p + a * (q - 0.5) * scale)
+                        .collect::<Vec<f64>>();
+                    // Add values to firefly position
+                    for s in 0..i.solution_mut().len() {
+                        x[u][s] += pos[s];
+                    }
```

Comment:
You could probably merge all of this (and the calculation of `rand`) into something like
```rust
izip!(individual[i].solution_mut(), individual[j].solution(), &scales)
    .map(|(xi, xj, scale)| { /* calculate rand, diff, pos, etc. */ (xi, pos) })
    .for_each(|(xi, &pos)| *xi += pos);
```

This saves allocating a vector for each `rand`, `diff`, and `pos` and hopefully makes the main formula somewhat more readable.

Directly after this should come the evaluation of `individual[i]`, because with the `.solution_mut()` we have reset its objective value.

Reply:
While it does say so in the Pseudocode, the (presumably) original code does not evaluate in the for-loop: https://de.mathworks.com/matlabcentral/fileexchange/29693-firefly-algorithm

Reply:
You're right, very weird. 

> However, I have some trouble implementing this properly, especially without cloning the population at least once.

And if we don't evaluate we also can't avoid cloning, so no choice here.


On the other hand, the [relatively popular Python implementation](https://github.com/firefly-cpp/FireflyAlgorithm/blob/fe6137c0065bc7066cf83bbe041256d1e5d6cba5/fireflyalgorithm/fireflyalgorithm.py#L32) also evaluates inside the loop.

Reply:
Thanks! Okay, I will probably have a deeper look into the algorithm again (and possibly also all others that I want to implement).

Maybe I will still do two variants for the FA, as it is obviously not quite clear how it is originally intended. Even if the evaluation is inside the loop(s), this is a bit strange and should result in a lot of function evaluations. Also, I would like an alternative that might be slightly different but fits better to MAHF's structure.